addons/l10n_vn_edi_viettel/models/sinvoice.py

```diff
@@ -0,0 +1,114 @@
+    _name = 'sinvoice.template'
+    _description = 'SInvoice template'
+
+    name = fields.Char(
+        string='Template Code',
+        required=True,
+    )
+    template_invoice_type = fields.Selection(
+        selection=[
+            # Circular 32 being deprecated, only display types according to the Circular 78
+            ('1', '1 - Value-added invoice'),
+            ('2', '2 - Sales invoice'),
+            ('3', '3 - Public assets sales'),
+            ('4', '4 - National reserve sales'),
+            ('5', '5 - Invoice for national reserve sales'),
+            ('6', '6 - Warehouse release note'),
+        ],
+        required=True,
+    )
+    invoice_symbols_ids = fields.One2many(
+        comodel_name='sinvoice.symbol',
+        inverse_name='invoice_template_id',
+    )
+
+    _sql_constraints = [
+        ('name_uniq', 'unique (name)', 'The template code must be unique!')
+    ]
+
+    @api.constrains('name', 'template_invoice_type')
+    def _constrains_changes(self):
+        """
+        Multiple API endpoints will use these data, we should thus not allow changing them if they have been used
+        for any invoices sent to sinvoice.
+        """
+        # The conditions are the same. If any symbols of the template are being used, we shouldn't allow editing it.
+        self.invoice_symbols_ids._constrains_changes()
+
+
+# Invoice symbol that needs to be passed to Sinvoice and will determine the prefix of the
+# invoice number on their system
+class SInvoiceSymbol(models.Model):
+    _name = 'sinvoice.symbol'
+    _description = 'SInvoice symbol'
+    """
+    The invoice symbols are made of multiple parts.
+    The symbols characters have meanings that will influence the resulting invoices:
+
+    The first character is either C (invoice with code from tax department) or K (without)
+    It is then followed by two number which represent the current year (24 in 2024)
+    Next, a single character representing additional invoice information:
+
+        T: E-invoices registered with tax authorities by enterprises, organizations or household businesses;
+        D: E-invoices used for sale of public property or national reserve goods or special e-invoices that do not have some mandatory contents of the “T” invoices;
+        L: E-invoices issued separately by tax authorities;
+        M: E-invoices generated by cash registers;
+        N: Electronic delivery and internal consignment notes;
+        B: Electronic delivery notes for goods sent to sales agents;
+        G: VAT invoices in the form of electronic stamps, tickets or cards;
+        H: Sales invoices in the form of electronic stamps, tickets or cards.
+
+    Finally, it ends with two characters that shall be decided by the seller for their management requirements.
+    """
+
+    name = fields.Char(
+        string='Symbol',
+        required=True,
+    )
+    invoice_template_id = fields.Many2one(
+        comodel_name='sinvoice.template',
```

Comment:
So you need one symbol per partner and a symbol is linked to a template. However, the default symbol is created automatically from the setting. Do we really need to create 2 new models for that? :thinking:

Reply:
One template can hold multiple symbols based on your needs.
You can see in the docstring how the symbols are defined.

We could make the template code and type fields on the symbol themselves if you prefer, but they are not the same data and it could lead to repeating the same values multiple times.
In the online platform, both of these are separate data and I think it will be clearer how to set them up if they are setup in a similar manner in Odoo.